### ISSUE TITLE
BackgroundService handles Exceptions getting thrown differently when tasks are not faulted vs faulted

### ISSUE THREAD

**Issue by PureKrome** [aspnet/Extensions#813]
I think how exceptions are handled with `BackgroundService` class is a bit confusing and may require some discussion.

- If an exception is thrown during startup (e.g. in Main) and before `ExecuteAsync` is handled, the app will crash.
- If an exception is thrown during `ExecuteAsync` _but before_ the first `await` is called, the app will crash.
- If an exception is thrown during `ExecuteAsync` _but during or after_ the first `await` is called, the app will ... semi-hang? It stop processing ... but ... Ctrl-C does successfully start the app gracefully terminating.

Now - I _hope_ i have the above steps/scenario's right -- I might have fudged something (async/await/ task contexts, etc are my weak points).

So the problem here is about the inconsistency about what an exception should be doing in a `BackgroundService`.  Should it crash the host (assuming no top level try/catch in Main)? or get swallowed? Maybe the issue is really about documentation (as the code is totally fine).

This issue arose when I started a conversation in the "ASP.NET slack" group when I was trying to handle an exception getting thrown in side the `ExecuteAsync` method and bubbling that upwards.

To quote some of the summaries of this convo:

> @davidfowl [7:06 PM]
> great discussion
> heh
> it could suppress both and  if you wanted to get exceptions, then you override StartAsync and put logic there

--

> @chrisvanderpennen [7:13 PM]
> theres imo three scenarios that need covering:
> i don't want any exceptions in this to stop my host
> i want an exception during init to stop the host startup, but runtime exceptions are ok
> i want a runtime exception to stop my host 

> @davidfowl [7:30 PM]
> somebody file a bug

so CC'ing @poke @gulbanana @chrisvanderpennen and @davidfowl who were in the convo.

- I think the issue subject might need to be refactored.
- happy to edit the main body of this issue, where i got my facts wrong.
- This issue needs the famous tag 'fowler' (fowler has issues). :)

edit: [this was a sample gist](https://gist.github.com/PureKrome/06421fa7170ab2ef348a0dd1be52d10b) I tried to make to also test/repo this problem.

**Comment by muratg** [comment]
@davidfowl  what's the work in 3.0 here?

cc  @Tratcher

**Comment by davidfowl** [comment]
A few targeted changes to BackgroundService. 
- We Task.Run the call to ExecuteAsync because user code cannot be trusted
- Exceptions thrown in the background service are observed in StopAsync but never in StartAsync.

**Comment by anurse** [comment]
@davidfowl are you going to do this?

**Comment by davidfowl** [comment]
Yes, I have a branch, I've just been lazy finishing it. I'll do it next week.

**Comment by aaronhudon-ts** [comment]
Curious if any has implemented a resilient pattern in their BackgroundService so that it will continue running even if an exception is thrown during ExecutionAsync

**Comment by PureKrome** [comment]
Sadness -> this is now in `Backlog` and will not ship with `v3`?

**Comment by anurse** [comment]
Correct. Unfortunately there was higher priority work that we had to bump this out for.

**Comment by chucker** [comment]
I would have expected worker services to have a built-in notion of global exception handling, but I guess not (yet)?

Anyway, here's what I do for now:

```csharp
        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            while (!stoppingToken.IsCancellationRequested)
            {
                try
                {
                    _logger.LogInformation("Worker running at: {time}", DateTimeOffset.Now);

                    // actual stuff happens here
                }
                catch (Exception ex)
                {
                    // waiting on https://github.com/aspnet/Extensions/issues/813
                    _Logger.LogError(ex, "Global exception occurred. Will resume in a moment.");
                }

                await Task.Delay(1000, stoppingToken);
            }
        }
```
